pytensor/gradient.py

```diff
@@ -2052,6 +2052,72 @@ def hessian(cost, wrt, consider_constant=None, disconnected_inputs="raise"):
     return as_list_or_tuple(using_list, using_tuple, hessians)
 
 
+def hessian_vector_product(cost, wrt, p, **grad_kwargs):
+    """Return the expression of the Hessian times a vector p.
+
+    Parameters
+    ----------
+    cost: Scalar (0-dimensional) variable.
+    wrt: Vector (1-dimensional tensor) 'Variable' or list of Vectors
+    p: Vector (1-dimensional tensor) 'Variable' or list of Vectors
+        Each vector will be used for the hessp wirt to exach input variable
+    **grad_kwargs:
+        Keyword arguments passed to `grad` function.
+
+    Returns
+    -------
+    :class:` Vector or list of Vectors
+        The Hessian times p of the `cost` with respect to (elements of) `wrt`.
+
+    Examples
+    --------
+
+    .. testcode::
+
+        import numpy as np
+        from scipy.optimize import minimize
+
+        from pytensor import function
+        from pytensor.tensor import vector
+        from pytensor.gradient import jacobian, hessian_vector_product
+
+        x = vector('x')
+        p = vector('p')
+
+        rosen = (100 * (x[1:] - x[:-1] ** 2) ** 2 + (1 - x[:-1]) ** 2).sum()
+        rosen_hessp = hessian_vector_product(rosen, x, p)
+        rosen_jac = jacobian(rosen, x)
+
+        rosen_fn = function([x], rosen)
+        rosen_jac_fn = function([x], rosen_jac)
+        rosen_hessp_fn = function([x, p], rosen_hessp)
+        x0 = np.array([1.3, 0.7, 0.8, 1.9, 1.2])
+        res = minimize(
+            rosen_fn,
+            x0,
+            method="Newton-CG",
+            jac=rosen_jac_fn,
+            hessp=rosen_hessp_fn,
+            options={"xatol": 1e-8},
+        )
+        assert res.success
+        np.testing.assert_allclose(res.x, np.ones_like(x0), atol=1e-3)
+
+
+    """
+    wrt_list = wrt if isinstance(wrt, Sequence) else [wrt]
+    p_list = p if isinstance(p, Sequence) else [p]
+    grad_wrt_list = grad(cost, wrt=wrt_list, **grad_kwargs)
+    Hp_list = [
+        grad((grad_wrt * p).sum(), wrt=wrt, **grad_kwargs)
+        for wrt, grad_wrt, p in zip(wrt_list, grad_wrt_list, p_list, strict=True)
+    ]
```

Comment:
I think this might be better:

```suggestion
    hessian_cost = sum([grad_wrt * p]).sum() for grad_wrt, p in zip(wrt_list, p_list, strict=True)
    Hp_list = pm.grad(hessian_cost, wrt=wrt_list, **grad_kwargs)
```

Reply:
Preference for sum on the outer loop vs add?

Reply:
I am not 100% sure, but I wouldn't be surprised if we end up with less duplicate code in the graph that the rewrites would have to (hopefully) clean up.

Reply:
I was not asking about the one grad call, I was asking hessian_cont = sum(terms) vs hessian_cost = add(*terms).

I did the later in my last push

Reply:
ah sorry. I think add is better?

Reply:
Added a note with a link to the docs in the docstring